pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
             None
         };
 
-        resources_state.pkgutil_modules_infos(py, prefix, state.optimize_level)
+        resources_state.pkgutil_modules_infos(py, prefix, state.optimize_level, |resource| {
+            _PathEntryFinder::is_visible("", &resource.name)
+        })
+    }
+
+    // Canonicalize the path to the current executable or raise an OSError.
+    //
+    // Even `std::env::current_exe()` can fail to `fs::canonicalize()` if, e.g.,
+    // the file is deleted after the program starts.
+    fn exe_realpath(&self, py: Python) -> PyResult<std::path::PathBuf> {
+        let current_exe = &self.state(py).get_resources_state().current_exe;
+        current_exe.canonicalize().map_err(|err| {
+            let exe = current_exe.display();
+            // Use a Python expression instead of PyErr::new to ensure we get the
+            // right OSError subclass.
+            let strerror = format!("cannot open current executable: '{}'", exe);
+            let raw_os_error = err
+                .raw_os_error()
+                .map_or_else(|| "None".to_string(), |err_code| err_code.to_string());
+            let code = if cfg!(windows) {
+                format!(
+                    "OSError(None, r\"{}\", r\"{}\", {})",
+                    strerror, exe, raw_os_error
+                )
+            } else {
+                format!("OSError({}, r\"{}\", r\"{}\")", raw_os_error, strerror, exe)
+            };
+            py.eval(&code, None, None).map_or_else(
+                |err_creating_exc| err_creating_exc,
+                |exc| PyErr::from_instance(py, exc),
+            )
+        })
```

Comment:
I just merged another PR to introduce `dunce::canonicalize()` to the `pyembed` crate. We should definitely use that crate to perform the path canonicalization so UNC paths don't enter the picture on Windows. `.current_exe` here may already be normalized, but I'd have to double check the code to be sure.

What I'm trying to say is that all this code can likely be deleted. At the worst we'd introduce another call to `dunce::canonicalize()`. Ideally `.current_exe` is already normalized on storage.

Reply:
> `.current_exe` here may already be normalized, but I'd have to double check the code to be sure.

[`dunce::canonicalize`] sets `ResolvedOxidizedPythonInterpreterConfig.current_exe` only if `OxidizedPythonInterpreterConfig.exe` wasn't already set and `OxidizedPythonInterpreterConfig::resolve` was the constructor:
https://github.com/indygreg/PyOxidizer/blob/05f4a67fe436728ee9d70d30768e1aa4c520ab9c/pyembed/src/config.rs#L282-L291

This might hold for [`MainPythonInterpreter::new`](https://github.com/indygreg/PyOxidizer/blob/05f4a67fe436728ee9d70d30768e1aa4c520ab9c/pyembed/src/interpreter.rs#L69) (again, if `config.exe` isn't set), but will not hold for `OxidizedImporter`s created from Python:
https://github.com/indygreg/PyOxidizer/blob/05f4a67fe436728ee9d70d30768e1aa4c520ab9c/pyembed/src/importer.rs#L949
which calls
https://github.com/indygreg/PyOxidizer/blob/05f4a67fe436728ee9d70d30768e1aa4c520ab9c/pyembed/src/python_resources.rs#L444

-----

> What I'm trying to say is that all this code can likely be deleted... Ideally `.current_exe` is already normalized on storage.

The ideal doesn't seem currently to match the code. I am eager to avoid bugs in whatever we merge into `main` from this PR, so I'd like to push back against deleting `exe_realpath` _for now_.

----

> At the worst we'd introduce another call to `dunce::canonicalize()`.

The surgery required to ensure [all locations that set `current_exe`](https://github.com/indygreg/PyOxidizer/blob/05f4a67fe436728ee9d70d30768e1aa4c520ab9c/pyembed/src/python_resources.rs#L395) are properly normalized strikes me as being beyond the scope of this PR. I count at least two different structs that own a `current_exe` (`OxidizedPythonInterpreterConfig` and `PythonResourcesState`), and both have a variety of constructors, not all of which appear to be used. Without narrowing down the places where `current_exe` gets set, we risk a bad or missing canonicalization sneaking back in in the future.

Instead I have two alternative proposals. I'd be okay with either one, so please let me know what you'd want to do.

1. Shore up the current code as is and merge it for now. I would create a ticket for fixing the `current_exe` canonicalization situation and the ticket would mention the opportunity for removing `OxidizedImporter::exe_realpath`.
2. Change the code to check for `os.join(sys.executable, "pkg", "mod")`. This would better match the documentation as currently written and would be easier for Python code to verify. It would differ from the way that embedded packages' `__path__`s are created, which is to prefix with `current_exe`. Note that `current_exe` does not have to be the same as `sys.executable`. In particular, `sys.executable` is writable from Python. Other reasons `current_exe` and `sys.executable` may not match include race conditions in reading symlinks during interpreter startup.

Until I hear back from you, I'm going to carry forward with shoring up the code as is. In particular, there's a bug in how the `OSError` is constructed that I'll fix.

[`dunce::canonicalize`]: https://docs.rs/dunce/1.0.1/dunce/fn.canonicalize.html

Reply:
That `OSError` construction bug is now fixed in 97096f1a03d49572c2077ce42f0f390369dae7e9